pkg/stream/xprotocol/factory.go

```diff
@@ -49,5 +50,23 @@ func (f *streamConnFactory) CreateBiDirectStream(context context.Context, connec
 }
 
 func (f *streamConnFactory) ProtocolMatch(context context.Context, prot string, magic []byte) error {
+	subProtocolMatchers := xprotocol.GetMatchers()
+	if subProtocolMatchers == nil {
+		return stream.FAILED
+	}
+	again := false
+	for _, matcher := range subProtocolMatchers {
+		result := matcher(magic)
+		if result == types.MatchSuccess {
+			return nil
+		}
+		if result == types.MatchAgain {
+			again = true
+		}
+	}
+	if again {
+		return stream.EAGAIN
```

Comment:
If the protocol is unknown, always return `stream.EAGAIN` will fall into dead loop.

Reply:
> If the protocol is unknown, always return `stream.EAGAIN` will fall into dead loop.

I don't understand. stream.EAGAIN only happened that subProtocolMatchers returns types.MatchAgain. if the protocol is unknown means each subProtocolMatchers will finally return types.MatchFailed. Then ProtocolMatch will return stream.FAILED